File: add_two_numbers.cpp
#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode* tail = new ListNode(0);
    ListNode* head = tail;


    int rem = 0;
    int s = 0;
    while (l1 || l2) {
        s = rem;
        if (l1) {
            s += l1->val;
            l1 = l1->next;
        }
        if (l2) {
            s += l2->val;
            l2 = l2->next;
        }
        if (s > 9) {
            rem = 1;
            s -= 10;
        }
        else {
            rem = 0;
        }
        tail->next = new ListNode(s);
        tail = tail->next;
    }

    if (rem) {
        tail->next = new ListNode(rem);
        tail = tail->next;
    }
    return head->next;
}

ListNode* construct_linked_list(const vector<int> &nums) {
    ListNode* tail = new ListNode();
    ListNode* head = tail;

    for (const int el: nums) {
        tail->next = new ListNode(el);
        tail = tail->next;
    }
    return head->next;
}

int main() {
    // ListNode* A = new ListNode(2);
    // A->next = new ListNode(4);
    // A->next->next = new ListNode(3);

    // ListNode* B = new ListNode(5);
    // B->next = new ListNode(6);
    // B->next->next = new ListNode(4);

    // auto ans = addTwoNumbers(A, B);
    // while (ans) {
    //     std::cout << ans->val << ' ';
    //     ans = ans->next;
    // }

    // ListNode* A = new ListNode(0);
    // ListNode* B = new ListNode(0);

    // auto ans = addTwoNumbers(A, B);
    // while (ans) {
    //     std::cout << ans->val << ' ';
    //     ans = ans->next;
    // }

    ListNode* A = construct_linked_list({9,9,9,9,9,9,9});
    ListNode* B = construct_linked_list({9,9,9,9});

    auto ans = addTwoNumbers(A, B);
    while (ans) {
        std::cout << ans->val << ' ';
        ans = ans->next;
    }

// l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]
// Output: [8,9,9,9,0,0,0,1]
}

File: python_solutions/add_two_numbers.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def addTwoNumbers(
    l1: Optional[ListNode],
    l2: Optional[ListNode]
)-> Optional[ListNode]:
    head = ListNode()
    ans = ListNode(next=head)

    rem = 0
    while l1 and l2:
        cur_sum = l1.val + l2.val + rem
        if cur_sum >= 10:
            rem = 1
            cur_sum -= 10
        else:
            rem = 0
        head.next = ListNode(cur_sum)
        head = head.next
        l1 = l1.next
        l2 = l2.next

    while l1:
        cur_sum = l1.val + rem
        if cur_sum == 10:
            rem = 1
            cur_sum = 0
        else:
            rem = 0
        head.next = ListNode(cur_sum)
        head = head.next
        l1 = l1.next

    while l2:
        cur_sum = l2.val + rem
        if cur_sum == 10:
            rem = 1
            cur_sum = 0
        else:
            rem = 0
        head.next = ListNode(cur_sum)
        head = head.next
        l2 = l2.next

    if rem == 1:
        head.next = ListNode(1)
        head = head.next

    return ans.next.next


def create_linked_list(nodes: List[ListNode]) -> ListNode:
    if len(nodes) == 1:
        return nodes[0]

    tmp = nodes[0]
    dummy = ListNode(next=tmp)
    for node in nodes[1:]:
        tmp.next = node
        tmp = tmp.next
    return dummy.next


def print_linked_list(node: ListNode):
    tmp = node
    while tmp:
        print(tmp.val)
        tmp = tmp.next


if __name__ == '__main__':
    # a = create_linked_list([ListNode(2), ListNode(4), ListNode(3)])
    # b = create_linked_list([ListNode(5), ListNode(6), ListNode(4)])

    # a = create_linked_list([ListNode(2), ListNode(4), ListNode(3), ListNode(7)])
    # b = create_linked_list([ListNode(5), ListNode(6), ListNode(4)])

    # a = create_linked_list([ListNode(9), ListNode(9), ListNode(9), ListNode(9), ListNode(9), ListNode(9), ListNode(9)])
    # b = create_linked_list([ListNode(9), ListNode(9), ListNode(9), ListNode(9)])

    a = create_linked_list([ListNode(0)])
    b = create_linked_list([ListNode(0)])

    ans = addTwoNumbers(a, b)

    print_linked_list(ans)